openTEPES/openTEPES.py

```diff
@@ -17,74 +17,75 @@
-    print(                 'Open Generation, Storage, and Transmission Operation and Expansion Planning Model with RES and ESS (openTEPES) - Version 4.17.8 - October 17, 2024', file=open(_path+'/openTEPES_version_'+CaseName+'.log','w'))
-
-    pIndOutputResults = [j for i,j in idxDict.items() if i == pIndOutputResults][0]
-    pIndLogConsole    = [j for i,j in idxDict.items() if i == pIndLogConsole   ][0]
-
-    # introduce cycle flow formulations in DC and AC load flow
-    pIndCycleFlow = 0
-
-    # Define sets and parameters
-    InputData(DirName, CaseName, mTEPES, pIndLogConsole)
-
-    # Define variables
-    SettingUpVariables(mTEPES, mTEPES)
-
-    # first/last stage
-    FirstST = 0
-    for st in mTEPES.st:
-        if FirstST == 0:
-            FirstST = 1
-            mTEPES.First_st = st
-        mTEPES.Last_st = st
-
-    # objective function and investment constraints
-    TotalObjectiveFunction    (mTEPES, mTEPES, pIndLogConsole)
-    InvestmentModelFormulation(mTEPES, mTEPES, pIndLogConsole)
-
-    # initialize parameter for dual variables
-    mTEPES.pDuals = {}
-
-    # control for not repeating the stages in case of several ones
-    mTEPES.NoRepetition = 0
-
-    # initialize the set of load levels up to the current stage
-    mTEPES.na = Set(initialize=[])
 
-    # iterative model formulation for each stage of a year
-    for p,sc,st in mTEPES.ps*mTEPES.stt:
+@dataclass
+class Simulation:
+    mTEPES: ConcreteModel
+    DirName: str
+    CaseName: str
+    SolverName: str
+    pIndOutputResults: int # TODO: make this bool instead
+    pIndLogConsole: int # TODO: make this bool instead
+    pIndCycleFlow: int = 0 # introduce cycle flow formulations in DC and AC load flow
+
+    def __post_init__(self):
+        #%% replacing string values by numerical values
+        idxDict        = dict()
+        idxDict[False] = 0
+        idxDict[0    ] = 0
+        idxDict[0.0  ] = 0
+        idxDict[0.0  ] = 0
+        idxDict['No' ] = 0
+        idxDict['NO' ] = 0
+        idxDict['no' ] = 0
+        idxDict['N'  ] = 0
+        idxDict['n'  ] = 0
+        idxDict[True ] = 1
+        idxDict['Yes'] = 1
+        idxDict['YES'] = 1
+        idxDict['yes'] = 1
+        idxDict['Y'  ] = 1
+        idxDict['y'  ] = 1
+
+        self.pIndLogConsole = idxDict[self.pIndLogConsole]
+        self.pIndOutputResults = idxDict[self.pIndOutputResults]
+        self.path = os.path.join(self.DirName, self.CaseName)
+
+    def input_data(self):
+        # Define sets and parameters
+        InputData(self.DirName, self.CaseName, self.mTEPES, self.pIndLogConsole)
+
+    def define_variables(self):
+        # Define variables
+        SettingUpVariables(self.mTEPES, self.mTEPES)
+
+    def setup_first_last_stage(self):
+        self.First_st = self.mTEPES.st.first()
+        self.Last_st = self.mTEPES.st.last()
```

Comment:
In addition to my previous comment, the first `st` should include the initial time step of the year when aggregating all time steps across stages. Similarly, the final `st` should represent the last time step. @arght, please let me know if there are any corrections needed.

Reply:
When I introduced mTEPES.First_st and mTEPES.Last_st was to allow periods where the first local levels do not correspond to the first hours of the year, which may happen if we assign duration 0 to the starting hours. For this reason I wrote these instructions
    # first/last stage
    FirstST = 0
    for st in mTEPES.st:
        if FirstST == 0:
            FirstST = 1
            mTEPES.First_st = st
        mTEPES.Last_st = st
instead of using .first() and .last() constructions from Pyomo. mTEPES.st only have the active stages (duration <>0) from the stage set.
I don't know if this clarifies the question.